### PR TITLE
Enable adding custom certificates to the Docker image

### PR DESCRIPTION
**Pull request by Martin-Idel-SI**
Enable adding custom certificates via `--build-arg` to a Docker build. The certificates will be added to the Java keystore as well as the system keystore like the proxy-certificates. This is necessary for certain company repositories which provide their own certificate authority and trip up the analyzer. 

Tested internally with repository having certificate problems.

### CODE REVIEW

File: docs/hints-for-use-with-docker.md [hunk 1]
```diff
@@ -14,6 +14,12 @@ docker run \
 
 **Note:** The single forward slash `/` between the environment variable `$PWD` and the `:` is required for PowerShell compatibility, as PowerShell otherwise interprets `:` as part of the environment variable. 
 
+### Setting custom certificates to Docker image keystore
+
+It is possible to install custom certificates when building the image so that they are installed into various keystores.
+To do this, specify `--build-arg CA_FILES=<path>` when running `docker build`. 
+Note that this requires the folder or certificate file to be inside the docker build context (usually the directory where the build is run, i.e. probably the directory the Dockerfile resides in). Otherwise, the folders cannot be copied into the Dockerfile.
```
Comment (MarcelBochtler) [inline docs/hints-for-use-with-docker.md:21]: Add a line break before the last sentence.
Also: I think you don't mean to say that the files cannot be copied into the Dockerfile, but into the docker image.

**Review by MarcelBochtler — APPROVED** [review]

File: scripts/import_certificates.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# License-Filename: LICENSE
+
+CERTIFICATE_FILES=$1
+
+if [ -z "$CERTIFICATE_FILES" ]; then
+  echo "No certificates specified. Skip installation."
```
Comment (sschuberth) [inline scripts/import_certificates.sh:23]: s/skip/skipping/
s/installation/import/ (as the script is called "import")

File: scripts/import_certificates.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  exit 0
+fi
+
+echo "Installing certificates $CERTIFICATE_FILES"
```
Comment (sschuberth) [inline scripts/import_certificates.sh:27]: s/Installing/Importing/

Either dot at end of sentence, or ":" after "certificates".

File: scripts/import_certificates.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# License-Filename: LICENSE
+
+CERTIFICATE_FILES=$1
```
Comment (sschuberth) [inline scripts/import_certificates.sh:20]: These aren't really the certificate files, but only the (single) prefix. They are resolved to files by appending the asterisk below. So why not continue to use `FILE_PREFIX` here, which also makes the diff smaller?

File: scripts/import_proxy_certs.sh [hunk 5]
```diff
@@ -60,31 +60,7 @@ fi
...
-    done
-else
-    echo "No JVM keystore found, skipping the import."
-fi
-
-# Also add the proxy certificates to the system certificates, e.g. for curl to work.
-echo "Adding proxy certificates to the system certificates..."
-cp $FILE_PREFIX* /usr/local/share/ca-certificates/
-update-ca-certificates
+DIR=$(dirname "$(readlink -f "$0")")
```
Comment (sschuberth) [inline scripts/import_proxy_certs.sh:63]: Better name: `SCRIPT_DIR`

File: Dockerfile [hunk 6]
```diff
@@ -117,8 +117,15 @@ RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/
 
 COPY --from=build /usr/local/src/ort/scripts/*.sh /opt/ort/bin/
 
+# This can be set to a folder containing crt-Files for custom certificates that ORT and all build tools should know about.
+ARG CA_FILES=""
```
Comment (sschuberth) [inline Dockerfile:121]: Are these really `CA` files, or `CRT` files?

File: Dockerfile [hunk 7]
```diff
@@ -117,8 +117,15 @@ RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/
 
 COPY --from=build /usr/local/src/ort/scripts/*.sh /opt/ort/bin/
 
+# This can be set to a folder containing crt-Files for custom certificates that ORT and all build tools should know about.
```
Comment (sschuberth) [inline Dockerfile:120]: Prefer "directory" over folder" (also in the commit message); the latter is so Windows-ish 😉

Comment (sschuberth) [reply Dockerfile:120]: s/crt-Files/CRT-files/

File: Dockerfile [hunk 8]
```diff
@@ -117,8 +117,15 @@ RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/
...
+ARG CA_FILES=""
+COPY "$CA_FILES" /tmp/certificates/
+
 # Custom install commands.
 RUN /opt/ort/bin/import_proxy_certs.sh && \
+    if [ -n "$CA_FILES" ]; then \
+      /opt/ort/bin/import_certificates.sh "/tmp/certificates/"; \
```
Comment (sschuberth) [inline Dockerfile:127]: Nit: No need to quote the argument here.

File: docs/hints-for-use-with-docker.md [hunk 9]
```diff
@@ -14,6 +14,13 @@ docker run \
 
 **Note:** The single forward slash `/` between the environment variable `$PWD` and the `:` is required for PowerShell compatibility, as PowerShell otherwise interprets `:` as part of the environment variable. 
 
+### Setting custom certificates to Docker image keystore
+
+It is possible to install custom certificates when building the image so that they are installed into various keystores.
+To do this, specify `--build-arg CA_FILES=<path>` when running `docker build`. 
+Note that this requires the folder or certificate file to be inside the docker build context (usually the directory where the build is run, i.e. probably the directory the Dockerfile resides in).
```
Comment (sschuberth) [inline docs/hints-for-use-with-docker.md:21]: s/folder/directory/
s/docker/Docker/

File: docs/hints-for-use-with-docker.md [hunk 10]
```diff
@@ -14,6 +14,13 @@ docker run \
 
 **Note:** The single forward slash `/` between the environment variable `$PWD` and the `:` is required for PowerShell compatibility, as PowerShell otherwise interprets `:` as part of the environment variable. 
 
+### Setting custom certificates to Docker image keystore
+
+It is possible to install custom certificates when building the image so that they are installed into various keystores.
+To do this, specify `--build-arg CA_FILES=<path>` when running `docker build`. 
+Note that this requires the folder or certificate file to be inside the docker build context (usually the directory where the build is run, i.e. probably the directory the Dockerfile resides in). 
+Otherwise the folders cannot be copied into the Docker image.
```
Comment (sschuberth) [inline docs/hints-for-use-with-docker.md:22]: s/folder/directory/

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 6 above)
Comment (Martin-Idel-SI) [reply Dockerfile:121]: For us those are exclusively CA files, but if you use self-signed certificates, you'd have to actually include the certificate itself...

File: scripts/import_certificates.sh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# License-Filename: LICENSE
+
+FILES_PREFIX=$1
```
Comment (sschuberth) [inline scripts/import_certificates.sh:20]: In the original script this is called `FILE_PREFIX` (not `FILES`) and so we should do here.

File: Dockerfile [hunk 12]
```diff
@@ -117,8 +117,15 @@ RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/
 
 COPY --from=build /usr/local/src/ort/scripts/*.sh /opt/ort/bin/
 
+# This can be set to a directory containing CRT-Files for custom certificates that ORT and all build tools should know about.
```
Comment (sschuberth) [inline Dockerfile:120]: I really meant `CRT-files` with lowercase "f".

The commit message still has "folder":

> Note that the folder has to be inside the Docker build context.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]

**Pull request merged**